enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,371 @@
+Plugins can use any Patternfly component. Only a single version of React and
+Patternfly will be loaded.
+
+Plugins will not be able to use legacy Patternfly 3 components.
+
+### Delivering Plugins
+
+Plugins are delivered through operators. The operators will typically create a
+deployment on the platform with an HTTP server to host the plugin and expose it
+using a Kubernetes `Service`. The `Service` must use HTTPS and a
+[service serving certificate](https://docs.openshift.com/container-platform/4.4/security/certificates/service-serving-certificate.html).
+The console backend will load the plugin directly from the `Service` using the
+service CA bundle.
+
+Operators declare that they have a console plugin available in the
+`ClusterServiceVersion` resource. This can be done through the
+`console.openshift.io/plugins` annotation on the CSV. This annotation is
+only a hint to console that a dynamic plugin is available. Console will not
+load the plugin until a cluster admin enables it.
+
+The value is a serialized JSON array of objects. Each object has a service
+name, port, and path to the plugin manifest.
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  annotations:
+    console.openshift.io/plugins: '[{ "service": "my-console-plugin", "port": 8443, "manifest": "/manifest.json" }]'
+```
+
+The namespace for the `Service` is assumed to be the same namespace as the
+`ClusterServiceVersion`. In the case of global operators, it will always be the
+namespace of the global `ClusterServiceVersion` (not a CSV copy).
+
+Plugins will be disabled by default. They will need to be manually enabled by a
+cluster administrator before console will load any plugin code. Console will
+provide a UI for enabling plugins. The list of enabled plugins is set in the
+console operator config.
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: Console
+metadata:
+  name: cluster
+spec:
+  managementState: Managed
+  plugins:
+  - service: my-console-plugin
+    namespace: openshift-operators
+    port: 8443
+    manifest: /manifest.json
+```
+
+This will trigger a new rollout of the console `Deployment` with the updated
+plugins. Note that this does mean the user will need to refresh the browser
+to see the new plugins.
+
+#### Alternative: ConsolePlugin CRD
+
+Rather than tying plugins to operators, we could have a CRD that allows you to
+declare a plugin.
+
+```yaml
+apiVersion: console.openshift.io/v1
+kind: ConsolePlugin
+metadata:
+  name: my-plugin
+spec:
+  displayName: My Plugin
+  service:
+    name: my-console-plugin
+    namespace: my-operator-namespace
+    port: 8443
+    manifest: /manifest.json
+```
+
+Even if we have the CRD, we'd still want the cluster admin to enable the plugin
+before it becomes active.
+
+A disadvantage to this approach is that it's harder to limit plugins to Red
+Hat operators. It's also harder to trace a plugin back to an operator that
+owns it.
```

Comment:
We went back and forth on this, and I left both options in to get feedback. The main motivation for using the annotation is we wanted it to be an internal API (at least to start).

Reply:
For an internal-ish API via CRDs, could you use v1alpha1 or some other clearly-unstable versioning?  I don't have a strong opinion either way, but "we are very likely to break this first attempt at an API" doesn't seem sufficient to rule out CRDs.

Pushing back on Ben's CVO-operators contributing, I am certainly not going to be as comfortable with console/UI/UX stuff as folks on the console team.  For components that are shipped with the core (and thus not vulnerable to the release-cycle issues given at the top of this enhancement), I expect moving to operator-controlled plugins would lead to a poorer-quality product, as folks like me fumbled around outside their wheelhouses.